src/app/icd/client/CheckInHandler.h

```diff
@@ -0,0 +1,146 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines objects for a CHIP check-in message unsolicited
+ *      handler
+ *
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandSender.h>
+#include <app/OperationalSessionSetup.h>
+#include <app/icd/client/CheckInDelegate.h>
+#include <app/icd/client/DefaultICDClientStorage.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+
+namespace chip {
+namespace app {
+
+class CheckInHandler : public Messaging::ExchangeDelegate, public Messaging::UnsolicitedMessageHandler
+{
+
+public:
+    CHIP_ERROR Init(Messaging::ExchangeManager * exchangeManager, ICDClientStorage * clientStorage, CheckInDelegate * delegate);
+    void Shutdown();
+
+    CheckInHandler();
+
+    virtual ~CheckInHandler() = default;
+
+    /**
+     * @brief Callback received on successfully establishing a CASE session in order to re-register the client with the peer node
+     * using a new key to avoid counter rollover problems.
+     *
+     * @param[in] context context of the client establishing the CASE session
+     * @param[in] exchangeMgr exchange manager to use for the re-registration
+     * @param[in] sessionHandle session handle to use for the re-registration
+     */
+    static void HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                      const SessionHandle & sessionHandle);
+    /**
+     * @brief Callback received on failure to establish a CASE session in order to re-register the client with the peer node using a
+     * new key to avoid counter rollover problems.
+     *
+     * @param[in] context context of the client establishing the CASE session
+     * @param[in] peerId Scoped Node ID of the peer node
+     * @param[in] err failure reason
+     */
+    static void HandleDeviceConnectionFailure(void * context, const ScopedNodeId & peerId, CHIP_ERROR err);
+
+    class RegisterCommandSenderCallback : public CommandSender::Callback
+    {
+    public:
+        void OnResponse(chip::app::CommandSender * apCommandSender, const chip::app::ConcreteCommandPath & aPath,
+                        const chip::app::StatusIB & aStatus, chip::TLV::TLVReader * aData) override
+        {}
+        void OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError) override { mError = aError; }
+        void OnDone(chip::app::CommandSender * apCommandSender) override {}
+
+        CHIP_ERROR mError = CHIP_NO_ERROR;
+    } registerCommandSenderDelegate;
```

Comment:
remove this, registerCommandSenderDelegate?

Reply:
Will be addressed in a follow-up PR with design changes.